### ISSUE TITLE
Upgrade to ethereum-contracts 1.4

### ISSUE THREAD

**Issue by d10r** [superfluid-finance/protocol-monorepo#1079]
Upgrade to [v1.4.0](https://github.com/superfluid-finance/protocol-monorepo/releases/tag/ethereum-contracts%40v1.4.0)

**Steps**
1. Run workflow
* Go to https://github.com/superfluid-finance/protocol-monorepo/actions, click "Deploy to Networks" (to be renamed to "Deploy to Mainnet")
* Click "Run workflow"
* set network name (canonical name)
* click "Run workflow"
* Wait for it to succeed
2. Verify newly deployed contracts
* Click the workflow, then the only job
* Expand the step "Upload new contract addresses"
* Locate the line saying "upload packages/ethereum-contracts/tmp/contract-addresses.txt to ..." and copy the string after "***" (to be replaced by the actual URL).
* append this string to "https://refs.superfluid.finance/", resulting in something like e.g. "https://refs.superfluid.finance/bsc-mainnet-contract-addrs@3082945142/contract-addresses.txt"
* Download that file, e.g. to "bsc.addrs"
* Run `tasks/etherscan-verify-framework.sh <network-name> <file>` with the file just downloaded as argument
* Wait for completion and make sure none of the verifications failed
3. Signing
* Notify signers by providing a link to an issue which links to workflow runs
* Signer: check the field "Data/Subject" in the list of Multisig transactions. Compare the list of contracts and their addresses with the output of the GH workflow run
* Signer: check that the newly deployed contracts are verified on the primary explorer
* Signer: sign
4. Execution
Can be done by the last signer or by proto-ops. Recommended gas limit for the tx: 5M

**Smoke testing items**
* new logic contracts are initialized

Workflow: https://github.com/superfluid-finance/protocol-monorepo/actions/workflows/handler.deploy-to-network.yml

Incremental improvements during the process: https://github.com/superfluid-finance/protocol-monorepo/pull/1076

# Testnets
Canary: avalanche-fuji: https://github.com/superfluid-finance/protocol-monorepo/runs/8196759405

All networks: https://github.com/superfluid-finance/protocol-monorepo/actions/runs/3009764543

# Mainnets
```
gh workflow run "Deploy to Network" -r protocol-deploy-refinements -f network=avalanche-c
```

## avalanche-c

- [x] governance upgrade

- [x] agreement forwarder

- [x] contracts upgrade

workflow run: https://github.com/superfluid-finance/protocol-monorepo/runs/8268551310?check_suite_focus=true
new contracts: https://refs.superfluid.finance/avalanche-c-contract-addrs@3021873773/contract-addresses.txt

## xdai-mainnet

- [x] governance upgrade

- [x] agreement forwarder

- [x] contracts upgrade
workflow run: https://github.com/superfluid-finance/protocol-monorepo/runs/8269432851?check_suite_focus=true
new contracts: https://refs.superfluid.finance/xdai-mainnet-contract-addrs@3022196873/contract-addresses.txt

## optimism-mainnet

- [x] governance upgrade

- [x] agreement forwarder

- [x] contracts upgrade
workflow run: https://github.com/superfluid-finance/protocol-monorepo/runs/8269968537?check_suite_focus=true
(1st failed attempts (RPC timeout): https://github.com/superfluid-finance/protocol-monorepo/runs/8269775779?check_suite_focus=true)
new contracts: https://refs.superfluid.finance/optimism-mainnet-contract-addrs@3022382135/contract-addresses.txt

## arbitrum-one

- [x] governance upgrade

- [x] agreement forwarder

- [x] contracts upgrade
workflow run: https://github.com/superfluid-finance/protocol-monorepo/runs/8271528464?check_suite_focus=true
(1st failed attempts (RPC timeout): https://github.com/superfluid-finance/protocol-monorepo/runs/8270576169?check_suite_focus=true)
new contracts: https://refs.superfluid.finance/arbitrum-one-contract-addrs@3022908024/contract-addresses.txt

## polygon-mainnet

- [x] governance upgrade

- [x] agreement forwarder

- [x] contracts upgrade
requires GAS_PRICE override in order to not fail with _transaction underpriced_
workflow run: https://github.com/superfluid-finance/protocol-monorepo/runs/8271884798?check_suite_focus=true
new contracts: https://refs.superfluid.finance/arbitrum-one-contract-addrs@3022908024/contract-addresses.txt

## BSC

- [x] governance upgrade

- [x] agreement forwarder

- [x] contracts upgrade
workflow run: https://github.com/superfluid-finance/protocol-monorepo/actions/runs/3082945142
new contracts: https://refs.superfluid.finance/bsc-mainnet-contract-addrs@3082945142/contract-addresses.txt

# Findings

Limitation: updated SuperToken logic can't be verified because it's deployed only once the gov upgrade is triggered (via gov.updateContracts() -> host.updateSuperTokenFactory() -> factory.updateCode() -> _updateSuperTokenLogic() -> createSuperTokenLogic() -> SuperTokenFactoryHelper.create()
(imo this way to deploy the contract is too convoluted and should be changed such that deployment happens before the gov action: That would also avoid attributing the gas costs to the gov tx)

**Comment by hellwolf** [comment]
Avalance C-Chain contracts upgraded: https://snowtrace.io/tx/0x567212ccaad53afe58b5e2169c710303c26771f9b3859f44e7d0407fce7b1322

**Comment by hellwolf** [comment]
Polygon contracts upgraded: https://polygonscan.com/tx/0x8a026e5d271bc79eb3c95134e41f1aa22942b4122da42f77985d9edc7f2af916

**Comment by hellwolf** [comment]
xDAI contracts upgraded: https://gnosisscan.io/tx/0x06d0034977fb2bdfdb81644d3c3a2b3f46fad6fec1222f5d3a47d09ca3ae3cea

**Comment by hellwolf** [comment]
Optimism contracts upgraded: https://optimistic.etherscan.io/tx/0xa502b11b5e874deec6e9946f29a460b761102b6eeafe3f04bec7ed28e7f9895f

**Comment by hellwolf** [comment]
Arbitrum contracts upgraded: https://arbiscan.io/tx/0xcddd1e923d8f12dc4124cd9ad6b6d23db113d4d22b0acb554a480ed609d4eb3e

**Comment by d10r** [comment]
new governance logic deployed on all networks and verified on all networks except Arbitrum (fails). Multisig tx to be confirmed and executed on all networks except Optimism and Avalanche (already done).

**Comment by d10r** [comment]
## CFAv1Forwarder

Deployed to 0xCfA1E187C9141B8bA90a436CB789017FA573d051 on all networks.

### testnets

```
for t in `cat testnets`; do echo $t && RELEASE_VERSION=v1 npx truffle exec --network $t scripts/deploy-deterministically.js : CFAv1Forwarder; done
for t in `cat testnets`; do echo $t && npx truffle run verify --network $t CFAv1Forwarder@0xcfa1e187c9141b8ba90a436cb789017fa573d051; done
for t in `cat testnets`; do echo $t && RELEASE_VERSION=v1 npx truffle exec --network $t scripts/gov-set-trusted-forwarder.js : 0x0000000000000000000000000000000000000000 0xCfA1E187C9141B8bA90a436CB789017FA573d051 1; done
```

Example tx creating a flow using `setFlowrate`: https://testnet.snowtrace.io/tx/0xcb283974a63b63a1a8e7b76e3dd317bf50b4e5236b3811a912bbf21f1b46632d


### mainnets

```
for m in `cat mainnets`; do echo $m && RELEASE_VERSION=v1 npx truffle exec --network $m scripts/deploy-deterministically.js : CFAv1Forwarder; done
# polygon needs special treatment
GAS_PRICE=44000000000 RELEASE_VERSION=v1 npx truffle exec --network polygon-mainnet scripts/deploy-deterministically.js : CFAv1Forwarder
# optimism needs special treatment
EST_TX_COST=702802300000000 RELEASE_VERSION=v1 npx truffle exec --network optimism-mainnet scripts/deploy-deterministically.js : CFAv1Forwarder

# verification
for m in `cat mainnets`; do echo $m && npx truffle run verify --network $m CFAv1Forwarder@0xcfa1e187c9141b8ba90a436cb789017fa573d051; done

# set as forwarder (prepare gov action)
for m in `cat mainnets`; do echo $m && POLYGON_MAINNET_GAS_PRICE=44e9 GOVERNANCE_ADMIN_TYPE=MULTISIG RELEASE_VERSION=v1 npx truffle exec --network $m scripts/gov-set-trusted-forwarder.js : 0x0000000000000000000000000000000000000000 0xCfA1E187C9141B8bA90a436CB789017FA573d051 1; done
```

**Comment by hellwolf** [comment]
BSC gov updated: https://bscscan.com/tx/0x60807c18fe71bf8bed005035da1fb1bb4e4535edeb20ec0c280ec583b2052ceb

BSC agreement forwarder trusted: https://bscscan.com/tx/0xb6fcb8025e1f645b6c6ca82cdd4419b1e948643bffb1335d79d02f09a1c767e6

**Comment by hellwolf** [comment]
Ava agreement forwarder trusted:

- https://snowtrace.io/tx/0x7270da75a6c6dbc8c42be82da2af805367e587cd377df8fd2e1abd6d9b0466bc
- https://snowtrace.io/address/0xcfa1e187c9141b8ba90a436cb789017fa573d051#writeContract

**Comment by hellwolf** [comment]
Gnosis Chain New Governance Upgraded:

- https://gnosisscan.io/address/0xcf3f4fa77f4427f6a1c1a96556194b40fc9dbf02#code
- approval: https://blockscout.com/xdai/mainnet/tx/0x0f6182ac28fcdf95c3d8d91695a800bbce1bbbb919baabbe713123fa282f13be

**Comment by hellwolf** [comment]
Gnosis Chain Trusted Forwarder Enabled:

- https://gnosisscan.io/address/0xcfa1e187c9141b8ba90a436cb789017fa573d051
- approval: https://blockscout.com/xdai/mainnet/tx/0x3d1151c3279c36a45b3a02715753b9193132ce6003d23d11bc036a2e560e9dd4

**Comment by hellwolf** [comment]
Polygon governance code upgrade:

- https://polygonscan.com/address/0xad0a83ef63a4cb62631c23bde5d52b29ab0bd6af#code
- approval: https://polygonscan.com/tx/0x8b42c9464ad0ede4ec13721681e88b014104fcf4123146f19c40088214d6a579

**Comment by hellwolf** [comment]
Polygon CFAv1Forwarder enabling trusted forwarder:

- https://polygonscan.com/address/0xcfa1e187c9141b8ba90a436cb789017fa573d051#code
- approval: https://polygonscan.com/tx/0xa9a49b553cb90b986d026bc0ea546d194ef768448b55d3c453600c4bb12e69f5

**Comment by hellwolf** [comment]
Optimism cfa agreement forwarder trusted forwarder:

- https://optimistic.etherscan.io/address/0xcfa1e187c9141b8ba90a436cb789017fa573d051#code
- approval: https://optimistic.etherscan.io/tx/0x9ec7d4f61a601abe20c85de688dbd37b0b998bce30d4b78f07884d91e97e5489

**Comment by hellwolf** [comment]
Arbitrum Governance Upgrade (source not verified, only bytecode matching) :

- https://arbiscan.io/address/0x87e00dced5670e01bee33a9a724b1dac790937ef 
- approval: https://arbiscan.io/tx/0x655c8b02c4b7fedb934f5393b3c1f6bf8d696598108aed5905ccdb3187c4efe6

**Comment by hellwolf** [comment]
Arbitrum cfa trusted forwarder trusted (source not verified, only bytecode matching) :

- https://arbiscan.io/address/0xcfa1e187c9141b8ba90a436cb789017fa573d051#code
- approval https://arbiscan.io/tx/0x58aafec42d28b776af045f1bed1023dabe33e53ff4812890d37c2e19c1d46b90
